File: core/src/nl/t64/game/rpg/screens/world/entity/PhysicsChest.java
package nl.t64.game.rpg.screens.world.entity;

import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.audio.AudioEvent;
import nl.t64.game.rpg.components.loot.Loot;
import nl.t64.game.rpg.components.party.HeroItem;
import nl.t64.game.rpg.components.party.SkillItemId;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.screens.world.entity.events.Event;
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent;
import nl.t64.game.rpg.screens.world.entity.events.SelectEvent;
import nl.t64.game.rpg.screens.world.entity.events.StateEvent;


public class PhysicsChest extends PhysicsComponent {

    private final Loot chest;
    private final StringBuilder stringBuilder;
    private boolean isSelected;

    public PhysicsChest(Loot chest) {
        this.chest = chest;
        this.stringBuilder = new StringBuilder();
        this.isSelected = false;
    }

    @Override
    public void receive(Event event) {
        if (event instanceof LoadEntityEvent loadEvent) {
            currentPosition = loadEvent.position;
            setBoundingBox();
        }
        if (event instanceof SelectEvent) {
            isSelected = true;
        }
    }

    @Override
    public void dispose() {
        // empty
    }

    @Override
    public void update(Entity entity, float dt) {
        if (isSelected) {
            isSelected = false;
            tryToOpenChest(entity);
        }
    }

    @Override
    void setBoundingBox() {
        boundingBox.set(currentPosition.x, currentPosition.y, Constant.TILE_SIZE, Constant.TILE_SIZE);
    }

    private void tryToOpenChest(Entity entity) {
        stringBuilder.setLength(0);
        if (isFailingOnTrap()) return;
        if (isFailingOnLock()) return;

        entity.send(new StateEvent(EntityState.OPENED));
        showFindDialog();
    }

    private boolean isFailingOnTrap() {
        if (chest.isTrapped()) {
            return isTrapToDifficult();
        }
        return false;
    }

    private boolean isTrapToDifficult() {
        HeroItem bestMechanic = Utils.getGameData().getParty().getHeroWithHighestSkill(SkillItemId.MECHANIC);
        if (canHandleTrapWith(bestMechanic)) {
            doHandleTrapWith(bestMechanic);
            return false;
        } else {
            dontHandleTrap();
            return true;
        }
    }

    private boolean canHandleTrapWith(HeroItem bestMechanic) {
        final int bestMechanicLevel = bestMechanic.getCalculatedTotalSkillOf(SkillItemId.MECHANIC);
        return chest.canDisarmTrap(bestMechanicLevel);
    }

    private void dontHandleTrap() {
        stringBuilder.append("There's a dangerous trap on this treasure chest.")
                     .append(System.lineSeparator())
                     .append(String.format("You need a level %s Mechanic to disarm the trap.", chest.getTrapLevel()));
        notifyShowMessageDialog(stringBuilder.toString());
    }

    private void doHandleTrapWith(HeroItem bestMechanic) {
        stringBuilder.append(String.format("%s disarmed the trap", bestMechanic.getName()));
        chest.disarmTrap();
    }

    private boolean isFailingOnLock() {
        if (chest.isLocked()) {
            return isLockToDifficult();
        }
        return false;
    }

    private boolean isLockToDifficult() {
        HeroItem bestThief = Utils.getGameData().getParty().getHeroWithHighestSkill(SkillItemId.THIEF);
        if (canHandleLockWith(bestThief)) {
            doHandleLockWith(bestThief);
            return false;
        } else {
            dontHandleLock();
            return true;
        }
    }

    private boolean canHandleLockWith(HeroItem bestThief) {
        final int bestThiefLevel = bestThief.getCalculatedTotalSkillOf(SkillItemId.THIEF);
        return chest.canPickLock(bestThiefLevel);
    }

    private void dontHandleLock() {
        if (stringBuilder.toString().isBlank()) {
            stringBuilder.append("There's a lock on this treasure chest.")
                         .append(System.lineSeparator());
        } else {
            stringBuilder.append(",")
                         .append(System.lineSeparator())
                         .append("but it seems the treasure chest is also locked.")
                         .append(System.lineSeparator());
        }
        stringBuilder.append(String.format("You need a level %s Thief to pick the lock.", chest.getLockLevel()));
        notifyShowMessageDialog(stringBuilder.toString());
    }

    private void doHandleLockWith(HeroItem bestThief) {
        if (stringBuilder.toString().contains(bestThief.getName())) {
            stringBuilder.append(System.lineSeparator())
                         .append("and picked the lock");
        } else if (stringBuilder.toString().isBlank()) {
            stringBuilder.append(String.format("%s picked the lock", bestThief.getName()));
        } else {
            stringBuilder.append(System.lineSeparator())
                         .append(String.format(" and %s picked the lock", bestThief.getName()));
        }
        chest.pickLock();
    }

    private void showFindDialog() {
        final String message = finishStringBuilder();
        if (message.isBlank()) {
            notifyShowFindDialog(chest, AudioEvent.SE_CHEST);
        } else {
            notifyShowFindDialog(chest, AudioEvent.SE_CHEST, message);
        }
    }

    private String finishStringBuilder() {
        if (!stringBuilder.toString().isBlank()) {
            stringBuilder.append('.');
        }
        return stringBuilder.toString();
    }

    @Override
    public void debug(ShapeRenderer shapeRenderer) {
        // empty
    }

}

File: core/src/nl/t64/game/rpg/screens/world/entity/PhysicsChest.kt
package nl.t64.game.rpg.screens.world.entity

import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.components.loot.Loot
import nl.t64.game.rpg.components.party.HeroItem
import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.world.entity.events.Event
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent
import nl.t64.game.rpg.screens.world.entity.events.OnActionEvent
import nl.t64.game.rpg.screens.world.entity.events.StateEvent


class PhysicsChest(private val chest: Loot) : PhysicsComponent() {

    private val stringBuilder: StringBuilder = StringBuilder()
    private var isSelected: Boolean = false

    override fun receive(event: Event) {
        if (event is LoadEntityEvent) {
            currentPosition = event.position
            setBoundingBox()
        }
        if (event is OnActionEvent) {
            if (event.playerDirection == Direction.NORTH
                && event.checkRect.overlaps(boundingBox)
            ) {
                isSelected = true
            }
        }
    }

    override fun update(entity: Entity, dt: Float) {
        if (isSelected) {
            isSelected = false
            tryToOpenChest(entity)
        }
    }

    override fun setBoundingBox() {
        boundingBox.set(currentPosition.x, currentPosition.y, Constant.TILE_SIZE, Constant.TILE_SIZE)
    }

    private fun tryToOpenChest(entity: Entity) {
        stringBuilder.setLength(0)
        if (isFailingOnTrap()) return
        if (isFailingOnLock()) return

        entity.send(StateEvent(EntityState.OPENED))
        showFindDialog()
    }

    private fun isFailingOnTrap(): Boolean {
        return if (chest.isTrapped()) isTrapToDifficult() else false
    }

    private fun isTrapToDifficult(): Boolean {
        val bestMechanic = gameData.party.getHeroWithHighestSkill(SkillItemId.MECHANIC)
        return if (canHandleTrapWith(bestMechanic)) {
            doHandleTrapWith(bestMechanic)
            false
        } else {
            dontHandleTrap()
            true
        }
    }

    private fun canHandleTrapWith(bestMechanic: HeroItem): Boolean {
        val bestMechanicLevel = bestMechanic.getCalculatedTotalSkillOf(SkillItemId.MECHANIC)
        return chest.canDisarmTrap(bestMechanicLevel)
    }

    private fun dontHandleTrap() {
        stringBuilder
            .append("There's a dangerous trap on this treasure chest.")
            .append(System.lineSeparator())
            .append("You need a level ${chest.trapLevel} Mechanic to disarm the trap.")
        brokerManager.componentObservers.notifyShowMessageDialog(stringBuilder.toString())
    }

    private fun doHandleTrapWith(bestMechanic: HeroItem) {
        stringBuilder.append("${bestMechanic.name} disarmed the trap")
        chest.disarmTrap()
    }

    private fun isFailingOnLock(): Boolean {
        return if (chest.isLocked()) isLockToDifficult() else false
    }

    private fun isLockToDifficult(): Boolean {
        val bestThief = gameData.party.getHeroWithHighestSkill(SkillItemId.THIEF)
        return if (canHandleLockWith(bestThief)) {
            doHandleLockWith(bestThief)
            false
        } else {
            dontHandleLock()
            true
        }
    }

    private fun canHandleLockWith(bestThief: HeroItem): Boolean {
        val bestThiefLevel = bestThief.getCalculatedTotalSkillOf(SkillItemId.THIEF)
        return chest.canPickLock(bestThiefLevel)
    }

    private fun dontHandleLock() {
        if (stringBuilder.toString().isBlank()) {
            stringBuilder
                .append("There's a lock on this treasure chest.")
                .append(System.lineSeparator())
        } else {
            stringBuilder
                .append(",")
                .append(System.lineSeparator())
                .append("but it seems the treasure chest is also locked.")
                .append(System.lineSeparator())
        }
        stringBuilder.append("You need a level ${chest.lockLevel} Thief to pick the lock.")
        brokerManager.componentObservers.notifyShowMessageDialog(stringBuilder.toString())
    }

    private fun doHandleLockWith(bestThief: HeroItem) {
        if (stringBuilder.toString().contains(bestThief.name)) {
            stringBuilder
                .append(System.lineSeparator())
                .append("and picked the lock")
        } else if (stringBuilder.toString().isBlank()) {
            stringBuilder
                .append("${bestThief.name} picked the lock")
        } else {
            stringBuilder
                .append(System.lineSeparator())
                .append(" and ${bestThief.name} picked the lock")
        }
        chest.pickLock()
    }

    private fun showFindDialog() {
        val message = finishStringBuilder()
        if (message.isBlank()) {
            brokerManager.componentObservers.notifyShowFindDialog(chest, AudioEvent.SE_CHEST)
        } else {
            brokerManager.componentObservers.notifyShowFindDialog(chest, AudioEvent.SE_CHEST, message)
        }
    }

    private fun finishStringBuilder(): String {
        if (stringBuilder.toString().isNotBlank()) {
            stringBuilder.append('.')
        }
        return stringBuilder.toString()
    }

    override fun debug(shapeRenderer: ShapeRenderer) {
        // empty
    }

}
